core/compiler.common.jvm/src/org/jetbrains/kotlin/load/java/AbstractAnnotationTypeQualifierResolver.kt

```diff
@@ -0,0 +1,205 @@
+
+    private fun Annotation.findAnnotation(fqName: FqName): Annotation? =
+        annotations.find { it.fqName == fqName }
+
+    private fun Annotation.hasAnnotation(fqName: FqName): Boolean =
+        annotations.any { it.fqName == fqName }
+
+    private val resolvedNicknames = ConcurrentHashMap<Any, Annotation>()
+
+    fun resolveTypeQualifierAnnotation(annotation: Annotation): Annotation? {
+        if (javaTypeEnhancementState.jsr305.isDisabled) return null
+        if (annotation.fqName in BUILT_IN_TYPE_QUALIFIER_FQ_NAMES || annotation.hasAnnotation(TYPE_QUALIFIER_FQNAME))
+            return annotation
+        if (!annotation.hasAnnotation(TYPE_QUALIFIER_NICKNAME_FQNAME))
+            return null
+        return resolvedNicknames.getOrPut(annotation.key) {
+            // This won't store nulls (ConcurrentHashMap does not permit that), but presumably unless the code
+            // is broken a nickname should be resolvable.
+            annotation.annotations.firstNotNullOfOrNull(::resolveTypeQualifierAnnotation) ?: return null
+        }
+    }
+
+    private fun resolveQualifierBuiltInDefaultAnnotation(annotation: Annotation): JavaDefaultQualifiers? {
+        if (javaTypeEnhancementState.disabledDefaultAnnotations) {
+            return null
+        }
+
+        return BUILT_IN_TYPE_QUALIFIER_DEFAULT_ANNOTATIONS[annotation.fqName]?.let { qualifierForDefaultingAnnotation ->
+            val state = resolveDefaultAnnotationState(annotation).takeIf { it != ReportLevel.IGNORE } ?: return null
+            qualifierForDefaultingAnnotation.copy(
+                nullabilityQualifier = qualifierForDefaultingAnnotation.nullabilityQualifier.copy(isForWarningOnly = state.isWarning)
+            )
+        }
+    }
+
+    private fun resolveDefaultAnnotationState(annotation: Annotation): ReportLevel {
+        val annotationFqname = annotation.fqName
+        if (annotationFqname != null && annotationFqname in JSPECIFY_DEFAULT_ANNOTATIONS) {
+            return javaTypeEnhancementState.getReportLevelForAnnotation(annotationFqname)
+        }
+        return resolveJsr305AnnotationState(annotation)
+    }
+
+    // We explicitly state that while JSR-305 TYPE_USE annotations effectively should be applied to every type.
+    // They are not applicable for type parameter bounds because it would be a breaking change otherwise.
+    private fun Set<AnnotationQualifierApplicabilityType>.allIfTypeUse(): Set<AnnotationQualifierApplicabilityType> =
+        if (AnnotationQualifierApplicabilityType.TYPE_USE in this)
+            AnnotationQualifierApplicabilityType.values().toSet() - AnnotationQualifierApplicabilityType.TYPE_PARAMETER_BOUNDS + this
+        else
+            this
+
+    private fun resolveTypeQualifierDefaultAnnotation(annotation: Annotation): TypeQualifierWithApplicability<Annotation>? {
+        if (javaTypeEnhancementState.jsr305.isDisabled) return null
```

Comment:
Is it used for jspecify? If so, we'd always return `null` if jsr305 is disabled. Otherwise, if it should be jsr305-specific, maybe rename this method?

Reply:
I kept the original name of this method, but now that it's private you can see it's only used in `extractDefaultQualifiers`, which does `resolveQualifierBuiltInDefaultAnnotation` first, and that's where jspecify is handled.

Reply:
Though doesn't the use of "TypeQualifierDefault" terminology imply JSR-305?

Reply:
> Though doesn't the use of "TypeQualifierDefault" terminology imply JSR-305?

Yes, it does, although someone else can perceive `TypeQualifierDefaultAnnotation` in a vacuum, referring it to any default annotation. Anyway I'd have single term for this in the compiler source code. `jsr305` itself is a good part of name IMHO to highlight jsr305 specific parts more explicitly.

Reply:
The separation of functions here is unfortunately not entirely clear, as AnnotationQualifiersFqNames.kt, JvmAnnotationNames.kt and JavaNullabilityAnnotationSettings.kt are in somewhat messy state, with duplicated and misplaced constants (e.g. why are JSPECIFY_DEFAULT_ANNOTATIONS part of BUILT_IN_TYPE_QUALIFIER_DEFAULT_ANNOTATIONS?) This should probably be refactored more extensively later.